subprojects/core/src/main/java/org/gradle/api/internal/artifacts/JavaEcosystemSupport.java

```diff
@@ -198,4 +200,74 @@ public void execute(CompatibilityCheckDetails<Usage> details) {
             }
         }
     }
+
+    @VisibleForTesting
+    public static class BundlingCompatibilityRules implements AttributeCompatibilityRule<Bundling>, ReusableAction {
+        private static final Set<String> COMPATIBLE_WITH_EXTERNAL = ImmutableSet.of(
+                // if we ask for "external" dependencies, it's still fine to bring a fat jar if nothing else is available
+                Bundling.EMBEDDED,
+                Bundling.SHADOWED
+        );
+
+        @Override
+        public void execute(CompatibilityCheckDetails<Bundling> details) {
+            Bundling consumerValue = details.getConsumerValue();
+            Bundling producerValue = details.getProducerValue();
+            if (consumerValue == null) {
+                // consumer didn't express any preference, everything fits
+                details.compatible();
+                return;
+            }
+            String consumerValueName = consumerValue.getName();
+            String producerValueName = producerValue.getName();
+            if (Bundling.EXTERNAL.equals(consumerValueName)) {
+                if (COMPATIBLE_WITH_EXTERNAL.contains(producerValueName)) {
```

Comment:
This indicates that `EXTERNAL` is not compatible with itself. Can it be an issue?
The call from `DefaultAttributesSchema` does not seem to care as equality has been verified before.
However in the call from `DependencyGraphBuilder#assertCompatibleAttributes` there is no prior equality test.

Reply:
This is because a compatibility rule is _never_ called when an attribute matches _exactly_.

Reply:
mmm, I see what you mean by `DependencyGraphBuilder#assertCompatibleAttributes`. Note that the Java compatibility rules do the same. I have to double check.

Reply:
Yeah I think we need to change `DependencyGraphBuilder` here.

Reply:
Actually we do: https://github.com/gradle/gradle/blob/166b30eded75ff98f7a7988bc1d251af172c898f/subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/builder/DependencyGraphBuilder.java#L419-L418

Reply:
Ah missed that part!

But I believe this is an area we could improve upon by explicitly documenting that contract in `CompatibilityCheckDetails` and making sure equality is always checked before.

Of course that's a topic for a subsequent PR.

Reply:
Created #8518